### PR TITLE
Adding `AnnDataFileInfo` and `AnnDataIngestParameters` to manage `AnnData` ingest jobs (SCP-4831)

### PR DESCRIPTION
**Pull request by bistline**
#### BACKGROUND
Since `AnnData` files can contain many data modalities - clustering, metadata, expression - we need a way of tracking what data has been extracted from an `AnnData` file.  Furthermore, we will need a mechanism for setting the various parameters for downstream ingest jobs once SCP file types have been extracted from a parent `AnnData` file.

#### CHANGES
This update adds two new models: `AnnDataFileInfo` and `AnnDataIngestParameters`.  `AnnDataFileInfo` is a small embedded class similar to `ExpressionFileInfo` that will track what data has been extracted from an `AnnData` file.  It also provides an interface for using the normal `Study`-bound methods for finding files of a given type, such as `study.expression_matrices`.  This ensures `AnnData` files that have had the corresponding data type extracted will be identified correctly.

`AnnDataIngestParameters` is analogous to the other `Parameterizable` classes that handle setting command line parameters for PAPI jobs, such as `DifferentialExpressionParameters`.  This will allow SCP to manage passing information from one job to the next without having to pull pipeline job metadata objects and parsing out command lines to get values for certain parameters.  Eventually, `AnnDataIngestParameters` will be able to handle all ingest-specific actions for `AnnData` files, from the initial "fragment" SCP file extracting to each downstream parse job for clustering, metadata, and expression.  Currently, it can handle extracting clustering files from original `h5ad` archives and launching `ingest_cluster` jobs once the extraction is complete.

#### MANUAL TESTING
1. Download https://cellxgene.cziscience.com/collections/2902f08c-f83c-470e-a541-e463e25e5058
2. Pull branch and boot all services to be able to upload and parse files (e.g. `bin/docker-compose-setup.sh`)
3. Sign into your local instance, go to the "Admin" interface and set the `Ingest Pipeline Docker Image` to `gcr.io/broad-singlecellportal-staging/scp-ingest-jb-anndata-delocalize-bugfix:896b506`
4. Go to the "Feature Flags" interface, and set the `ingest_anndata_file` flag for your email to `true`
5. Create a new study, and upload `local.h5ad` as an `AnnData` file, and confirm that once the file uploads a job is launch to perform the initial extraction
6. After 1-2 minutes, you should receive an email saying that the `AnnData file ingest has completed`
7. In `development.log`, you should see two new jobs launched to ingest the extracted clustering files for `X_tsne` and `X_umap`:
```
Remote found for local.h5ad:638e2ea4ace7570f36a9ff0e, launching Ingest job
2022-12-05 12:53:14 -0500: Deleting caches for SCP81
2022-12-05 12:53:14 -0500: caches successfully cleared
Request object sent to Google Pipelines API (PAPI), excluding 'environment' parameters:
---
:actions:
  :commands:
  - python
  - ingest_pipeline.py
  - "--study-id"
  - 63878fb7d17c6300547fe22b
  - "--study-file-id"
  - 638e2ea4ace7570f36a9ff0e
  - "--user-metrics-uuid"
  - f775823b-94f7-46af-b1c7-789582efaa18
  - ingest_cluster
  - "--ingest-cluster"
  - "--cluster-file"
  - gs://fc-45f2ff6d-92f5-4cad-9c6b-a4a832c2a2d7/_scp_internal/anndata_ingest/X_tsne.cluster.anndata_segment.tsv
  - "--name"
  - X_tsne
  - "--domain-ranges"
  - "{}"
  :flags: []
  :image_uri: gcr.io/broad-singlecellportal-staging/scp-ingest-jb-anndata-delocalize-bugfix:896b506
  :labels: {}
  :timeout: {}
```
8. After another 1-2 minutes, you should receive success emails for the two extracted clustering files
9. Go to the "Details" page for your study, and confirm that you see two new cluster objects with the following info:
Screen Shot 2022-12-05 at 1 29 24 PM

### CODE REVIEW

**Review by ehanna4 — COMMENTED** [review]
Going through the testing steps all worked as advertized however, the AnnData file doesn't appear to be available in the AnnData section nor in the download tab (see screenshot), is this expected? I would think we'd still want the full AnnData file in the Upload Wizard in the AnnData section and the download tab but that might be part of other work so just figured I'd to ask

**Comment by bistline** [comment]
> Going through the testing steps all worked as advertized however, the AnnData file doesn't appear to be available in the AnnData section nor in the download tab (see screenshot), is this expected? I would think we'd still want the full AnnData file in the Upload Wizard in the AnnData section and the download tab but that might be part of other work so just figured I'd to ask

No that is not expected!  Great catch.  Let me check out what's going on there.

**Comment by ehanna4** [comment]
@bistline okay was able to get things working, I made a new study and uploaded the file and can see the clusters and the file.

The very first time I tried before the file name was `local (1).h5ad` which automatically was converted to `local_(1)_.h5ad` in the wizard and that failed, then I retried the same file in the same study after renaming said file to `local1.h5ad` which produced that first screenshot. So must have found a weird edge case, but the happy path does work!

Screen Shot 2022-12-07 at 4 35 13 PM

**Review by ehanna4 — APPROVED** [review]
functional review - works as expected
code review - changes look good to me! Excited at the simplicity of this approach compared to the way I had been thinking about this.

File: app/models/ann_data_ingest_parameters.rb [hunk 1]
```diff
@@ -0,0 +1,66 @@
+# handles launching ingest jobs for AnnData files and derived SCP file fragments
+# usage patterns:
+# extract SCP files: AnnDataIngestParameters.new(anndata_file: study_file.gs_url)
+# parse extracted cluster file: AnnDataIngestParameters.new(
+#   ingest_anndata: false, extract_cluster: false, name: 'X_tsne', ingest_cluster: '--flag-only', domain_ranges: "{}"
```
Comment (eweitz) [inline app/models/ann_data_ingest_parameters.rb:5]: `--flag-only` seems unnecessarily unintuitive.  After reading, I understand it's used to construct a CLI argument in flag format -- i.e., only passing the name of the argument as an implied "true"; no argument value is passed.

I think it'd be a lot more understandable at a glance, and better encapsulated, to make such parameters in Ruby simple Booleans.  Using a specific string for effectively `true` while using a Boolean for `false` is unintuitive.  I can see why it was done this way, I just think it requires maintainers to build too much mental context.  

In other words, I think the benefit this particular mapping of the Ruby method interface to that Python command interface is outweighed by the impeded code understandability of mixing strings and Booleans for these parameter values.  It's too tightly coupled.  

All Boolean arguments in the Ingest Pipeline CLI have this form, so any argument that has `true` in Ruby bindings for the Ingest Pipeline CLI can be serialized as a flag-only argument.  Simple Booleans seems easier for maintainers to understand at a glance.

Comment (eweitz) [inline app/models/ann_data_ingest_parameters.rb:5]:
```suggestion
#   ingest_anndata: false, extract_cluster: false, name: 'X_tsne', ingest_cluster: true, domain_ranges: "{}"
```

File: app/models/ann_data_ingest_parameters.rb [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+  validates :anndata_file, :cluster_file,
+            format: { with: Parameterizable::GS_URL_REGEXP, message: 'is not a valid GS url' },
+            allow_blank: true
+
+  # default values for parameters
+  # '--flag-only' attributes are passed to the command line as a standalone flag with no value, e.g. --extract-cluster
+  # any parameters that are set to nil/false will not be passed to the command line
+  PARAM_DEFAULTS = {
+    ingest_anndata: '--flag-only',
+    extract_cluster: '--flag-only',
```
Comment (eweitz) [inline app/models/ann_data_ingest_parameters.rb:32]:
```suggestion
    ingest_anndata: true,
    extract_cluster: true,
```

File: app/models/concerns/parameterizable.rb [hunk 3]
```diff
@@ -34,7 +34,12 @@ def self.to_cli_opt(param_name)
   def to_options_array
     options_array = []
     attributes.each do |attr_name, value|
-      options_array += [Parameterizable.to_cli_opt(attr_name), value.to_s] if value.present?
+      # handling of flag-only options that have no value, like --extract-cluster
+      if value.to_s == '--flag-only'
```
Comment (eweitz) [inline app/models/concerns/parameterizable.rb:38]:
```suggestion
      if value == true
```

File: app/models/ingest_job.rb [hunk 4]
```diff
@@ -543,6 +579,31 @@ def set_has_image_cache
...
+    study_file.save
+  end
+
+  # launch appropriate downstream jobs once an AnnData file successfully extracts "fragment" files
+  def launch_anndata_subparse_jobs
+    if params_object.extract_cluster.present?
+      params_object.attribute_as_array(:obsm_keys).each do |fragment|
+        cluster_gs_url = params_object.fragment_file_gs_url(study.bucket_id, 'cluster', fragment)
+        cluster_params = AnnDataIngestParameters.new(
+          ingest_cluster: '--flag-only', name: fragment, cluster_file: cluster_gs_url, domain_ranges: '{}',
```
Comment (eweitz) [inline app/models/ingest_job.rb:598]:
```suggestion
          ingest_cluster: true, name: fragment, cluster_file: cluster_gs_url, domain_ranges: '{}',
```

File: test/models/ann_data_ingest_parameters_test.rb [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+  before(:all) do
+    @extract_params = {
+      anndata_file: 'gs://test_bucket/test.h5ad',
+    }
+
+    @ingest_cluster_params = {
+      ingest_anndata: false,
+      extract_cluster: false,
+      obsm_keys: nil,
+      ingest_cluster: '--flag-only',
```
Comment (eweitz) [inline test/models/ann_data_ingest_parameters_test.rb:14]:
```suggestion
      ingest_cluster: true,
```

File: test/models/ann_data_ingest_parameters_test.rb [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+      name: 'X_umap',
+      domain_ranges: '{}'
+    }
+  end
+
+  test 'should instantiate and validate params' do
+    extraction = AnnDataIngestParameters.new(@extract_params)
+    assert extraction.valid?
+    %i[ingest_anndata extract_cluster].each do |attr|
+      assert_equal '--flag-only', extraction.send(attr)
```
Comment (eweitz) [inline test/models/ann_data_ingest_parameters_test.rb:25]:
```suggestion
      assert_equal true, extraction.send(attr)
```

File: test/models/ann_data_ingest_parameters_test.rb [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+    end
+    %i[ingest_cluster cluster_file name domain_ranges].each do |attr|
+      assert extraction.send(attr).blank?
+    end
+    cmd = "--ingest-anndata --anndata-file gs://test_bucket/test.h5ad --extract-cluster --obsm-keys ['X_umap','X_tsne']"
+    assert_equal cmd, extraction.to_options_array.join(' ')
+    cluster_ingest = AnnDataIngestParameters.new(@ingest_cluster_params)
+    assert cluster_ingest.valid?
+    assert_equal '--flag-only', cluster_ingest.ingest_cluster
```
Comment (eweitz) [inline test/models/ann_data_ingest_parameters_test.rb:34]:
```suggestion
    assert_equal true, cluster_ingest.ingest_cluster
```

**Review by eweitz — APPROVED** [review]
Code looks broadly good!  I suggest a non-blocking maintainability improvement.

File: app/models/ann_data_ingest_parameters.rb (hunk 1 above)
Comment (bistline) [reply app/models/ann_data_ingest_parameters.rb:5]: That makes total sense - I'll figure out a way to simplify that interface.  Agreed it is non-intuitive.

File: test/models/ann_data_ingest_parameters_test.rb (hunk 6 above)
Comment (bistline) [reply test/models/ann_data_ingest_parameters_test.rb]: [c457c29](https://github.com/broadinstitute/single_cell_portal_core/pull/1689/commits/c457c29f9c13e6a4f6ff87723a3b5d44b778c0e7)

File: test/models/ann_data_ingest_parameters_test.rb (hunk 5 above)
Comment (bistline) [reply test/models/ann_data_ingest_parameters_test.rb]: [c457c29](https://github.com/broadinstitute/single_cell_portal_core/pull/1689/commits/c457c29f9c13e6a4f6ff87723a3b5d44b778c0e7)

File: app/models/ingest_job.rb (hunk 4 above)
Comment (bistline) [reply app/models/ingest_job.rb]: [c457c29](https://github.com/broadinstitute/single_cell_portal_core/pull/1689/commits/c457c29f9c13e6a4f6ff87723a3b5d44b778c0e7)

File: app/models/concerns/parameterizable.rb (hunk 3 above)
Comment (bistline) [reply app/models/concerns/parameterizable.rb]: [c457c29](https://github.com/broadinstitute/single_cell_portal_core/pull/1689/commits/c457c29f9c13e6a4f6ff87723a3b5d44b778c0e7)

File: app/models/ann_data_ingest_parameters.rb (hunk 2 above)
Comment (bistline) [reply app/models/ann_data_ingest_parameters.rb]: [c457c29](https://github.com/broadinstitute/single_cell_portal_core/pull/1689/commits/c457c29f9c13e6a4f6ff87723a3b5d44b778c0e7)

File: app/models/ann_data_ingest_parameters.rb (hunk 1 above)
Comment (bistline) [reply app/models/ann_data_ingest_parameters.rb]: [c457c29](https://github.com/broadinstitute/single_cell_portal_core/pull/1689/commits/c457c29f9c13e6a4f6ff87723a3b5d44b778c0e7)

Comment (bistline) [reply app/models/ann_data_ingest_parameters.rb]: [c457c29](https://github.com/broadinstitute/single_cell_portal_core/pull/1689/commits/c457c29f9c13e6a4f6ff87723a3b5d44b778c0e7)

File: test/models/ann_data_ingest_parameters_test.rb (hunk 7 above)
Comment (bistline) [reply test/models/ann_data_ingest_parameters_test.rb]: [c457c29](https://github.com/broadinstitute/single_cell_portal_core/pull/1689/commits/c457c29f9c13e6a4f6ff87723a3b5d44b778c0e7)

**Comment by bistline** [comment]
The last two builds have failed with the same error in different places: `RuntimeError: Wrapped undumpable exception for: RuntimeError` for some upstream call to orchestration.  Marking this as a false positive on the build failure.
